src/Components/Web.Extensions/src/InputFile/RemoteFileListEntryStream.cs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Web.Extensions
+{
+    // This class streams data from JS within the existing API limits of IJSRuntime.
+    // To produce good throughput, it prefetches up to buffer_size data from JS
+    // even when the consumer isn't asking for that much data, and does so by making
+    // N parallel requests in parallel (N ~= buffer_size / max_message_size).
+    //
+    // This should be understood as a TEMPORARY way to achieve the desired API and
+    // reasonable performance. Longer term we can surely replace this with something
+    // simpler and cleaner, either:
+    //
+    //  - Extending JS interop to allow streaming responses via SignalR's built-in
+    //    binary streaming support. That should reduce all of this to triviality.
+    //  - Or, failing that, at least use something like System.IO.Pipelines to manage
+    //    the supply/consumption of byte data with less custom code.
+
+    internal class RemoteFileListEntryStream : FileListEntryStream
```

Comment:
@javiercn You spoke before about doing a security review on the new file upload feature. I don't know if you had something else in mind, but one thing that occurs to me is reviewing this kind of logic to see if we can see any way it could be misused to make the server allocate arbitrary amounts of memory, or to leak memory, to get into an infinite loop, or anything like that.

Do you think that's something we should do directly within this PR review, or schedule something separate (maybe with Levi giving it a look over)?